pinot-transport/src/main/java/com/linkedin/pinot/routing/builder/LargeClusterRoutingTableBuilder.java

```diff
@@ -0,0 +1,299 @@
+          String state = instanceAndState.getValue();
+
+          // Only consider partitions that are ONLINE
+          if (!CommonConstants.Helix.StateModel.SegmentOnlineOfflineStateModel.ONLINE.equals(state)) {
+            continue;
+          }
+
+          // Skip instances that are disabled
+          if (pruner.isInactive(instance)) {
+            continue;
+          }
+
+          // Add to the instance -> segments map
+          Set<String> segmentsForInstance = instanceToSegmentMap.get(instance);
+
+          if (segmentsForInstance == null) {
+            segmentsForInstance = new HashSet<>();
+            instanceToSegmentMap.put(instance, segmentsForInstance);
+          }
+
+          segmentsForInstance.add(segment);
+
+          // Add to the segment -> instances map
+          instancesForSegment.add(instance);
+
+          // Add to the valid segments map
+          segmentsWithAtLeastOneOnlineReplica.add(segment);
+        }
+      }
+
+      allValidInstancesSet = instanceToSegmentMap.keySet();
+      instanceArray = allValidInstancesSet.toArray(new String[allValidInstancesSet.size()]);
+
+      segmentToInstanceArrayMap = new HashMap<>();
+    }
+
+    private Map<String, Set<String>> generateRoutingTable() {
+      // List of segments that have no instance serving them
+      Set<String> segmentsNotHandledByServers = new HashSet<>(segmentsWithAtLeastOneOnlineReplica);
+
+      // List of servers in this routing table
+      Set<String> instancesInRoutingTable = new HashSet<>(MAX_SERVERS_PER_QUERY);
+
+      // If there are not enough instances, add them all
+      if (instanceArray.length <= MAX_SERVERS_PER_QUERY) {
+        instancesInRoutingTable.addAll(allValidInstancesSet);
+        segmentsNotHandledByServers.clear();
+      } else {
+        // Otherwise add MAX_SERVERS_PER_QUERY instances
+        while (instancesInRoutingTable.size() < MAX_SERVERS_PER_QUERY) {
+          String randomInstance = instanceArray[random.nextInt(instanceArray.length)];
+          instancesInRoutingTable.add(randomInstance);
+          segmentsNotHandledByServers.removeAll(instanceToSegmentMap.get(randomInstance));
+        }
+      }
+
+      // If there are segments that have no instance that can serve them, add a server to serve them
+      while (!segmentsNotHandledByServers.isEmpty()) {
+        // Get the instances in array format
+        String segmentNotHandledByServers = segmentsNotHandledByServers.iterator().next();
+
+        String[] instancesArrayForThisSegment = segmentToInstanceArrayMap.get(segmentNotHandledByServers);
+
+        if (instancesArrayForThisSegment == null) {
+          Set<String> instanceSet = segmentToInstanceMap.get(segmentNotHandledByServers);
+          instancesArrayForThisSegment = instanceSet.toArray(new String[instanceSet.size()]);
+          segmentToInstanceArrayMap.put(segmentNotHandledByServers, instancesArrayForThisSegment);
+        }
+
+        // Pick a random instance
```

Comment:
I would extend the comment to include "and add all segments served by that instance" to the instances in routing table